crates/next-api/src/module_graph.rs

```diff
@@ -285,33 +285,63 @@ impl ClientReferencesGraph {
+
+            // Track how we reached each client reference.  This way if a client reference is
+            // referenced by the root and by a server component we don't only associate it with the
+            // server component.
+            #[derive(PartialEq, Eq, Copy, Clone)]
+            enum ParentType {
+                ServerComponent,
+                Page,
+                Both,
+            }
+            impl ParentType {
+                fn merge(left: Self, right: Self) -> Self {
+                    if left == right {
+                        left
+                    } else {
+                        // One is Both or one is ServerComponent and the other is Page, which means
+                        // Both
+                        Self::Both
+                    }
+                }
+            }
+            // Perform a DFS traversal to collect all client references and the set of server
+            // components for each module.
             graph.traverse_edges_from_entries_dfs(
                 entries,
-                // state_map is `module -> Option< the current so parent server component >`
+                // state_map is `module -> ParentType` for whether the parent is in a
                 &mut FxHashMap::default(),
                 |parent_info, node, state_map| {
                     let module = node.module();
-                    let module_type = data.get(&module);
-
-                    let current_server_component = if let Some(
-                        ClientReferenceMapType::ServerComponent(module),
-                    ) = module_type
-                    {
-                        Some(*module)
-                    } else if let Some((parent_node, _)) = parent_info {
-                        *state_map.get(&parent_node.module).unwrap()
-                    } else {
-                        // a root node
-                        None
-                    };
-
-                    state_map.insert(module, current_server_component);
+                    let module_type = data.client_references.get(&module);
+
+                    let parent_type =
+                        if let Some(ClientReferenceMapType::ServerComponent(_)) = module_type {
+                            ParentType::ServerComponent
+                        } else if let Some((parent_node, _)) = parent_info {
+                            *state_map.get(&parent_node.module).unwrap()
+                        } else {
+                            // a root node
+                            ParentType::Page
+                        };
+
+                    match state_map.entry(module) {
+                        Entry::Occupied(mut occupied_entry) => {
+                            let merged = ParentType::merge(*occupied_entry.get(), parent_type);
+                            if merged != parent_type {
+                                occupied_entry.insert(merged);
+                            }
+                        }
```

Comment:
Is it even possible for this to be occupied already? This is a DFS with no revisiting.